confdgnmi/src/confd_gnmi_api_adapter.py

```diff
@@ -142,52 +172,96 @@ def iter(kp, op, oldv, newv, state):
+                # op1\nxpath1\nval1\nop2\nxpath2\nval2 .....
+                # op1 .. first operation1, xpath1 .... first xpath, ...
+                # op is string used in ChangeOp Enum class
+                # currently operation can be only "modified"
+                # the size must be smaller then size in recv
+                data = msg.decode().split('\n')
+                assert len(data) % 3 == 0
+                chunks = [data[x:x + 3] for x in range(0, len(data), 3)]
+                self._append_changes(chunks)
+                self.put_event(self.SubscriptionEvent.SEND_CHANGES)
+                log.debug("data=%s", data)
+            log.info("<==")
+
+        def process_changes(self, external_changes=False):
             log.debug("==>")
             # make subscription for all self.monitored_paths
             with socket() as sub_sock:
                 prio = 10
                 cdb.connect(sub_sock, cdb.SUBSCRIPTION_SOCKET, '127.0.0.1',
                             _confd.CONFD_PORT)
+                found_in_cdb = has_non_cdb = False
                 for p in self.monitored_paths:
                     log.debug("subscribing config p=%s", p)
                     # TODO hash - breaks generic usage
                     # TODO for now we subscribe path for both, config and oper,
                     # TODO subscribe only for paths that exist
                     # it may be more efficient to find out type of path and subscribe
                     # only for one type
-                    cdb.subscribe2(sub_sock, cdb.SUB_RUNNING, 0, prio,
-                                   0, p)
-                    log.debug("subscribing operational p=%s", p)
-                    cdb.subscribe2(sub_sock, cdb.SUB_OPERATIONAL, 0, prio,
-                                   0, p)
-                cdb.subscribe_done(sub_sock)
+                    cs_node = _confd.cs_node_cd(None, p)
+                    is_cdb = cs_node.info().flags() & _confd.CS_NODE_IS_CDB
+                    if is_cdb:
+                        found_in_cdb = True
+                        cdb.subscribe2(sub_sock, cdb.SUB_RUNNING, 0, prio,
+                                       0, p)
+                        log.debug("subscribing operational p=%s", p)
+                        cdb.subscribe2(sub_sock, cdb.SUB_OPERATIONAL, 0, prio,
+                                       0, p)
+                    else:
+                        has_non_cdb = True
+                if found_in_cdb:
+                    cdb.subscribe_done(sub_sock)
                 log.debug("subscribe_done")
                 assert self.stop_pipe is not None
                 rlist = [sub_sock, self.stop_pipe[0]]
                 wlist = elist = []
                 try:
+                    ext_server_sock = None
+                    if external_changes and has_non_cdb:
+                        log.info("Starting external change server!")
+                        ext_server_sock = socket()
```

Comment:
One more candidate to refactoring: with small changes you can do `with socket() as ext_server_sock:...` instead of this and the `final` clause with `if`: just make the `while` loop a separate method.